apps/email/js/html_cache_restore.js

```diff
@@ -102,103 +250,236 @@ window.htmlCacheRestorePendingMessage = [];
+  window.globalOnAppMessage = function(listener) {
 
-  function startApp() {
-    var scriptNode = document.createElement('script');
+    Object.keys(listener).forEach(function(key) {
+      var fn = handlers[key] = listener[key];
+      var queue = handlerQueues[key];
+      if (queue.length) {
+        handlerQueues[key] = [];
+        queue.forEach(function(argsArray) {
+          fn.apply(undefined, argsArray);
+        });
+      }
+    });
 
-    if (loader) {
-      scriptNode.setAttribute('data-main', loadSrc);
-      scriptNode.src = loader;
+    // Only need to do this wiring once, but globalOnAppMessage could be called
+    // multiple times.
+    if (!evt) {
+      require(['evt'], function(ev) {
+        evt = ev;
+        evt.on('notification', onNotification);
+      });
+    }
+
+    return startupData;
+  };
+
+  // Attach the hasAccount so that other code can use it and always get the
+  // freshest cached value.
+  window.globalOnAppMessage.hasAccount = hasAccount;
+
+  function dispatch(type, args) {
+    if (handlers[type]) {
+      return handlers[type].apply(undefined, args);
     } else {
-      scriptNode.src = loadSrc;
+      handlerQueues[type].push(args);
     }
 
-    document.head.appendChild(scriptNode);
+    // On the very first dispatch when app open is triggered by a dispatchable
+    // event, need to finish bootup now that full startup state is known.
+    finishStartup();
   }
 
-  // TODO: mozHasPendingMessage can only be called once?
-  // Need to set up variable to delay normal code logic later
-  if (navigator.mozHasPendingMessage) {
-    if (navigator.mozHasPendingMessage('activity')) {
-      window.htmlCacheRestorePendingMessage.push('activity');
+  /**
+   * Perform requested activity.
+   *
+   * @param {MozActivityRequestHandler} req activity invocation.
+   */
+  function onActivityRequest(req) {
+    console.log('mozSetMessageHandler: received an activity');
+    if (messageTooFast('activity') || ignoreMessage('activity')) {
+      return req.postError('cancelled');
     }
-    if (navigator.mozHasPendingMessage('request-sync')) {
-      window.htmlCacheRestorePendingMessage.push('request-sync');
+
+    // Right now all activity entry points go to compose, but may need to
+    // revisited if the activity entry points change.
+    if (startingUp) {
+      startupData.view = 'compose';
+      hydrateHtml(startupData.view);
     }
-    if (navigator.mozHasPendingMessage('notification')) {
-      window.htmlCacheRestorePendingMessage.push('notification');
+
+    dispatch('activity', [req]);
+  }
+
+  function onNotification(msg) {
+    console.log('mozSetMessageHandler: received a notification');
+    // Skip notification events that are not from a notification "click". The
+    // system app will also notify this method of any close events for
+    // notifications, which are not at all interesting.
+    if (!msg.clicked) {
+      if (startingUp) {
+        window.close();
```

Comment:
(That is, infer from the fact that pendingMessage did not get nulled out that we didn't find a useful 'notification' activity and therefore the 'notification' event was a lie, and since we clearly had nothing better to do, we should now shutdown.  I guess based on my request-sync improvement, though, you'd need to check that flag too.  You could also have a boolean like "doingSomething" and just set that to true anyplace we start something.)